contracts/0.8.25/utils/Confirmations.sol

```diff
@@ -0,0 +1,217 @@
+    }
+
+
+    /**
+     * @notice Returns the confirmation expiry.
+     * @return The confirmation expiry in seconds.
+     */
+    function getConfirmExpiry() public view returns (uint256) {
+        return _getConfirmationsStorage().confirmExpiry;
+    }
+
+    /**
+     * @notice Returns the confirmation expiry for a given call data and confirmer.
+     * @param _callData The call data of the function.
+     * @param _role The role of the confirmer.
+     * @return The confirmation expiry in seconds.
+     */
+    function confirmations(bytes memory _callData, bytes32 _role) external view returns (uint256) {
+        return _getConfirmationsStorage().confirmations[_callData][_role];
+    }
+
+    /**
+     * @dev Restricts execution of the function unless confirmed by all specified roles.
+     * Confirmation, in this context, is a call to the same function with the same arguments.
+     *
+     * The confirmation process works as follows:
+     * 1. When a role member calls the function:
+     *    - Their confirmation is counted immediately
+     *    - If not enough confirmations exist, their confirmation is recorded
+     *    - If they're not a member of any of the specified roles, the call reverts
+     *
+     * 2. Confirmation counting:
+     *    - Counts the current caller's confirmations if they're a member of any of the specified roles
+     *    - Counts existing confirmations that are not expired, i.e. expiry is not exceeded
+     *
+     * 3. Execution:
+     *    - If all members of the specified roles have confirmed, executes the function
+     *    - On successful execution, clears all confirmations for this call
+     *    - If not enough confirmations, stores the current confirmations
+     *    - Thus, if the caller has all the roles, the function is executed immediately
+     *
+     * 4. Gas Optimization:
+     *    - Confirmations are stored in a deferred manner using a memory array
+     *    - Confirmation storage writes only occur if the function cannot be executed immediately
+     *    - This prevents unnecessary storage writes when all confirmations are present,
+     *      because the confirmations are cleared anyway after the function is executed,
+     *    - i.e. this optimization is beneficial for the deciding caller and
+     *      saves 1 storage write for each role the deciding caller has
+     *
+     * @param _calldata msg.data of the call (selector + arguments)
+     * @param _roles Array of role identifiers that must confirm the call in order to execute it
+     * @return bool True if the call was executed, false otherwise
+     *
+     * @notice Confirmations past their expiry are not counted and must be recast
+     * @notice Only members of the specified roles can submit confirmations
+     * @notice The order of confirmations does not matter
+     *
+     */
+    function _checkConfirmations(bytes calldata _calldata, bytes32[] memory _roles) internal returns (bool) {
```

Comment:
```suggestion
    function _collectConfirmation(bytes calldata _calldata, bytes32[] memory _roles) internal returns (bool) {
```

Reply:
🤔 this is one-off function that does not allow to check confirmations after the quorum was gathered. But the name implies like it does